### PR TITLE
Assignment 5

### PR DESCRIPTION
**Pull request by ybaghlaf**
just starting the assignment

### CODE REVIEW

File: SecretSanta/src/SecretSanta.Api/Controllers/UsersController.cs [hunk 1]
```diff
@@ -53,6 +61,8 @@ public ActionResult Delete(int id)
         }
 
         [HttpPut("{id}")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
```
Comment (Keboo) [inline SecretSanta/src/SecretSanta.Api/Controllers/UsersController.cs:65]: This endpoint also returns NoFound (404) in addition to these two.

File: SecretSanta/src/SecretSanta.Web/Controllers/UsersController.cs [hunk 2]
```diff
@@ -38,18 +60,18 @@ public IActionResult Edit(UserViewModel viewModel)
         {
             if (ModelState.IsValid)
             {
-                MockData.Users[viewModel.Id] = viewModel;
+                //MockData.Users[viewModel.Id] = viewModel;
```
Comment (Keboo) [inline SecretSanta/src/SecretSanta.Web/Controllers/UsersController.cs:63]: This should also be invoking the IUserClient Post endpoint to update the the user.

File: SecretSanta/tests/SecretSanta.Web.Tests/Controllers/UsersControllerTest.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using SecretSanta.Web.Api;
+using SecretSanta.Web.Tests.Api;
+using SecretSanta.Web.ViewModels;
+
+namespace SecretSanta.Web.Tests
+{
+    [TestClass]
+    public class UsersControllerTest
```
Comment (Keboo) [inline SecretSanta/tests/SecretSanta.Web.Tests/Controllers/UsersControllerTest.cs:13]: This is a good start, but this only tests two of the end points. The Delete and Edit endpoints should also be tested.

File: SecretSanta/tests/SecretSanta.Web.Tests/SecretSanta.Web.Tests.csproj [hunk 4]
```diff
@@ -0,0 +1,17 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (Keboo) [inline SecretSanta/tests/SecretSanta.Web.Tests/SecretSanta.Web.Tests.csproj:1]: This is close, but it is in a "tests" namespace, but the target directory name should be "test" not "tests"

**Review by Keboo — COMMENTED** [review]
- Update the API project with the following
  - Add a new Dto namespace ❌
  - The `UserController` should only accept or return classes from the Dto namespace. Create any needed classes there and update the controller to use them. You will need to convert between the DTO classes and the relevant entities as needed. ❌
  - Add OpenAPI (swagger) endpoint ✔
  - Use the OpenAPI to generate a C# client with relevant interfaces ✔
  - Decorate all relevant API endpoints in the `UserController` with the appropriate [ProducesResponseType](https://docs.microsoft.com/en-us/aspnet/core/web-api/action-return-types?view=aspnetcore-5.0#iactionresult-type) ✔
- Update the Web project with the following
  - Register the generated `UserClients` class using its interface with the Dependency Injection (DI) service. ✔
  - Update the `UserController` to take in the generated client as a dependency. ✔
  - Convert the `UserController` from using the `MockData` to using the client to get data from the API service. ❌ - Edit does not post using service
  - Update the controller to convert back and forth between the ViewModel classes and the API DTO classes as needed. ✔
- Update the SecretSanta.Api.Tests with the following:
  - The controller tests should be [interation tests](https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-5.0) not unit tests. ✔
  - All endpoints should be tested. ❌ - only index and create are tested
  - The `UserController` should use a test double for the `IUserRepository` to avoid needing to retest business and data logic. ❌
- Create a new SecretSanta.Web.Tests projecs with the following:
  - It should be inside of the test directory. ✔ - It was named "tests" not "test"
  - It should have nullable reference types enabled. ✔
  - It should have functional tests for all endpoints in the `UserController` that access the API. ❌ - Only Index and Create are tested
  - It should use a test double for the API service. This test double should be used to validate the controller's interaction with the API without actually calling the API. ✔

**Comment by ybaghlaf** [comment]
I think I missed that one Kevin! Whoops

**Pull request closed without merging**